### ISSUE TITLE
Is it possible to use this tool without [CLS] token?

### ISSUE THREAD

**Issue by zeyuyun1** [hila-chefer/Transformer-Explainability#16]
Hi, 

Thanks for the great tool. Say if I have a regression task (predicting some score given texts), I was wondering if it's possible to use this tool if I don't have [CLS] token in my input sequence?

Thank you!

**Comment by hila-chefer** [comment]
Hi @zeyuyun1 thanks for your interest in our work!

could you please be a bit more specific on which architecture you’re using? If I understand correctly your input is a sequence of text tokens, but how is the output calculated? Is the architecture based on pure self attention? Is it a transformer?

**Comment by zeyuyun1** [comment]
Sure. First, the input is a sequence of text tokens. The architecture is BERT so transformer and pure attention.

How the output the calculated is a little bit complicated. The detail is described in a recent work I published: [paper](https://arxiv.org/pdf/2103.15949.pdf) [demo](https://transformervis.github.io/transformervis/). You can see in the last section of the paper, I use the LIME algorithm to generate the saliency map, which is super super inefficient, which is why I am really interested in your method. If you are interested in the idea in the paper maybe we can collaborate. 

Here's how I calculate the output:

Given a sequence of text tokens, the hidden output of BERT is a sequence of vectors that can be considered the contextualized embedding. We pick one of the vectors and do sparse coding inference. 

For example, the output of ["I","love","NLP"] from 10th layer of BERT model is [v_1,v_2,v_3]. Let's say we are interested in v_2. Then we do argmin_a ||v_2- Wa||^2 + ||a||_1

W is a dictionary we learned (the detail of how we get it is from the paper above). We are trying to find a sparse code "a" that explains v_2 the best. Finding "a" is a LASSO problem.

Finally, we use a specific entry of "a" as our output. 

So given input X, BERT as a neural network f( ), Dictionary W, Our pipeline is:
 
X ----> f(X)[i] ----> "a"[j]. The output is a scalar.

**Comment by hila-chefer** [comment]
Thanks for clarifying @zeyuyun1, and congrats on your fascinating new work being released!
Firstly, I suggest you use the new version of our method as proposed in our new paper, you can see the code [here](https://github.com/hila-chefer/Transformer-MM-Explainability). The difference is that we remove LRP for simplicity, and it may be significantly easier to use.
Secondly, given that all your calculations are differentiable, all you need to do is propagate gradients from your network’s output back to each self-attention layer.
Once you do that, assuming you wish to visualize an intermediate layer, say layer 10, you can use our method and back propagate from the last layer only until the 10th layer. You can see the start_layer variable in our rollout functions.
Our method should provide you with a txt matrix where t is the number of tokens. The i-th row scores all the tokens for the i-th token, i.e. scores how much each token is significant to the i-th token. If I understand what you wrote correctly, this is what you wish to do for v_2 which is actually the second token (second row of our output matrix).
I would recommend to discard the influence of the i-th token on itself (i.e. the diagonal of the matrix) which is influenced by the identity matrix we add. This too is addressed in our second paper.
Seems to me that given your description, it is possible to use our method, but if I misunderstood anything feel free to correct me.
Also, please feel free to email me at hilachefer@mail.tau.ac.il or reach out to me via any social media you see fit, I would be happy to help if I can.
Good luck with your brand new work!

**Comment by hila-chefer** [comment]
Closing due to inactivity, @zeyuyun1 please reopen if necessary.

**Comment by zeyuyun1** [comment]
Thank you so much for your detailed reply! You can leave it closed for now. 
I need to think about it for a little bit more if the i-th row score is what I want. It is certainly the most intuitive way to think about it for now. However, the hidden state sequence might not represent position information anymore, especially in deeper layers.

**Comment by hila-chefer** [comment]
No problem, please note that while I agree that the tokens are contextualized, our method considers that using the rollout component, i.e. we consider the mixing of tokens in all previous layers. Given what you reminded me of, I correct my initial response- your start layer should be 0 and you should use rollout until the layer you are interested in. In our mock example it’s layer 10.
The rolling out component allows us to consider all the previous layers which mixed the tokens, so the matrix corresponding to the 10th layer actually considers *all* layers until the 10th layer.
The i-th layer scores represents the scores after the contextualization of all previous layers. You can think of our method as 2 folded- the first is constructing a relevancy matrix txt per layer, and the second is aggregating the layers to consider the fact that each layer mixes the tokens.

**Comment by zeyuyun1** [comment]
Hi hila,

That actually clarified my confusion. I think you are totally right. The i-th row score should be what I want. The only difference between my task and the classification task is that I am using i'th position of the hidden state sequence instead of the 0'th position (where the classification token is at).

Thank you again for the great tool! I really appreciate your help and quick reply.

**Comment by hila-chefer** [comment]
Happy to help! We added support for CLIP in our [new repo](https://github.com/hila-chefer/Transformer-MM-Explainability), I think the implementation there may be helpful since the output is also a score and not logits :) hope this helps
